lib/SILGen/SILGenPoly.cpp

```diff
@@ -3760,52 +3709,43 @@ SILGenModule::getOrCreateAutoDiffAssociatedFunctionThunk(
   auto assocFnRefType = assocFnRef->getType().castTo<SILFunctionType>();
 
   // Collect thunk arguments, converting ownership.
-  SmallVector<AllocStackInst *, 8> localAllocations;
-  SmallVector<SILValue, 8> argumentsToFree;
   SmallVector<SILValue, 8> arguments;
   for (auto *indRes : indirectResults)
     arguments.push_back(indRes);
-  forwardFunctionArgumentsConvertingOwnership(
-      thunkSGF, loc, assocFnRefType, origAssocFnType, params, arguments,
-      localAllocations, argumentsToFree);
+  forwardFunctionArguments(thunkSGF, loc, assocFnRefType, params, arguments);
   // Apply function argument.
   auto apply = thunkSGF.emitApplyWithRethrow(
       loc, assocFnRef, /*substFnType*/ assocFnRef->getType(),
       thunk->getForwardingSubstitutionMap(), arguments);
 
-  SmallVector<SILValue, 8> directResults;
-  extractAllElements(apply, loc, thunkSGF.B, directResults);
-  auto linearMap = ManagedValue::forUnmanaged(directResults.back());
-  auto linearMapFnType = linearMap.getType().castTo<SILFunctionType>();
-  auto targetLinearMapFnType = thunk->mapTypeIntoContext(
-      origAssocFnType->getResults().back().getSILStorageType())
-          .castTo<SILFunctionType>();
-
   // Create return instruction in the thunk, first deallocating local
   // allocations and freeing arguments-to-free.
   auto createReturn = [&](SILValue retValue) {
-    // Free arguments-to-free.
-    for (auto arg : argumentsToFree)
-      if (arg->getType().isObject())
-        thunkSGF.B.createReleaseValue(loc, arg,
-                                      thunkSGF.B.getDefaultAtomicity());
-      else
-        thunkSGF.B.createDestroyAddr(loc, arg);
-    // Deallocate local allocations.
-    for (auto *alloc : reversed(localAllocations))
-      thunkSGF.B.createDeallocStack(loc, alloc);
+    // Emit cleanups.
+    thunkSGF.Cleanups.emitCleanupsForReturn(
+        CleanupLocation::get(loc), NotForUnwind);
     // Create return.
     thunkSGF.B.createReturn(loc, retValue);
   };
 
   // If self ordering is not necessary and linear map types are unchanged,
   // return the `apply` instruction.
+  auto linearMapFnType = cast<SILFunctionType>(
+      thunk->mapTypeIntoContext(assocFnRefType->getResults().back().getType())
+          ->getCanonicalType());
+  auto targetLinearMapFnType = thunk->mapTypeIntoContext(
+      origAssocFnType->getResults().back().getSILStorageType())
+          .castTo<SILFunctionType>();
   if (!reorderSelf && linearMapFnType == targetLinearMapFnType) {
     createReturn(apply);
     return thunk;
   }
 
   // Otherwise, apply reabstraction/self reordering thunk to linear map.
+  SmallVector<SILValue, 8> directResults;
+  extractAllElements(apply, loc, thunkSGF.B, directResults);
+  auto linearMap = ManagedValue::forBorrowedObjectRValue(directResults.back());
```

Comment:
This cannot be a borrowed rvalue. It is owned because it's a function call's result.

Reply:
That makes sense, but I'm not sure what API to use instead. What do you recommend?

`linearMap` is known to be always consumed by `partial_apply`, so no cleanup is necessary.

We could use `ManagedValue::forOwnedObjectRValue` (which takes a `CleanupHandle`) and then disable the cleanup, but I'm not sure how to construct a `CleanupHandle`.

Reply:
Changed to `SILGenFunction::emitManagedRValueWithCleanup` instead in https://github.com/apple/swift/pull/27393/commits/6b770a32de1e385359175322981e6b381e8dd3ef. Functionally identically here but more semantically accurate.

Reply:
`emitManagedRValueWithCleanup` is the right API to use.